src/Voron/StorageEnvironment.cs

```diff
@@ -174,33 +178,50 @@ public StorageEnvironment(StorageEnvironmentOptions options)
                 throw;
             }
         }
-        private async Task IdleFlushTimer()
+
+        ~StorageEnvironment()
+        {
+            if (_log.IsOperationsEnabled)
+                _log.Operations($"Finalizer of storage environment was called. Env: {this}");
+
+            Dispose();
+        }
+
+        private static async Task IdleFlushTimer(EnvironmentWeakReference weakRef)
         {
             try
             {
-                var cancellationToken = _cancellationTokenSource.Token;
-
-                while (cancellationToken.IsCancellationRequested == false)
+                while (true)
                 {
-                    if (Disposed)
+                    if (weakRef.IsAlive == false)
+                        break;
+
+                    // IMPORTANT: we're intentionally accessing the env via weakRef.Environment instead of creating local variable to avoid copying it to
+                    // the state-machine produced by the async method (it's is a struct with fields corresponding to the locals in the source method)
+                    // otherwise a reference to the environment is hold what prevents from running the finalizer of the storage environment
+
```

Comment:
How do we prevent local promotion here?

Reply:
What do you mean by "local promotion"? You mean why it's not copied to the async state?